bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogManagerForEntryLogPerLedger.java

```diff
@@ -0,0 +1,463 @@
+        this.entryLogAndLockTupleCacheLoader = new CacheLoader<Long, EntryLogAndLockTuple>() {
+            @Override
+            public EntryLogAndLockTuple load(Long key) throws Exception {
+                return new EntryLogAndLockTuple();
+            }
+        };
+        /*
+         * Currently we are relying on access time based eviction policy for
+         * removal of EntryLogAndLockTuple, so if the EntryLogAndLockTuple of
+         * the ledger is not accessed in
+         * entrylogMapAccessExpiryTimeInSeconds period, it will be removed
+         * from the cache.
+         *
+         * We are going to introduce explicit advisory writeClose call, with
+         * that explicit call EntryLogAndLockTuple of the ledger will be
+         * removed from the cache. But still timebased eviciton policy is
+         * needed because it is not guaranteed that Bookie/EntryLogger would
+         * receive successfully write close call in all the cases.
+         */
+        ledgerIdEntryLogMap = CacheBuilder.newBuilder()
+                .expireAfterAccess(entrylogMapAccessExpiryTimeInSeconds, TimeUnit.SECONDS)
+                .maximumSize(maximumNumberOfActiveEntryLogs)
+                .removalListener(new RemovalListener<Long, EntryLogAndLockTuple>() {
+                    @Override
+                    public void onRemoval(
+                            RemovalNotification<Long, EntryLogAndLockTuple> expiredLedgerEntryLogMapEntry) {
+                        removalOnExpiry(expiredLedgerEntryLogMapEntry);
+                    }
+                }).build(entryLogAndLockTupleCacheLoader);
+    }
+
+    /*
+     * This method is called when access time of that ledger has elapsed
+     * entrylogMapAccessExpiryTimeInSeconds period and the entry for that
+     * ledger is removed from cache. Since the entrylog of this ledger is
+     * not active anymore it has to be removed from
+     * replicaOfCurrentLogChannels and added to rotatedLogChannels.
+     *
+     * Because of performance/optimizations concerns the cleanup maintenance
+     * operations wont happen automatically, for more info on eviction
+     * cleanup maintenance tasks -
+     * https://google.github.io/guava/releases/19.0/api/docs/com/google/
+     * common/cache/CacheBuilder.html
+     *
+     */
+    private void removalOnExpiry(RemovalNotification<Long, EntryLogAndLockTuple> expiredLedgerEntryLogMapEntry) {
+        Long ledgerId = expiredLedgerEntryLogMapEntry.getKey();
+        log.debug("LedgerId {} is not accessed for entrylogMapAccessExpiryTimeInSeconds"
+                + " period so it is being evicted from the cache map", ledgerId);
+        EntryLogAndLockTuple entryLogAndLockTuple = expiredLedgerEntryLogMapEntry.getValue();
+        Lock lock = entryLogAndLockTuple.ledgerLock;
+        BufferedLogChannel logChannel = entryLogAndLockTuple.entryLog;
+        lock.lock();
+        try {
+            replicaOfCurrentLogChannels.remove(logChannel.getLogId());
```

Comment:
just one question:

is the eviction behavior same, no matter of the `RemovalCause`?

http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/cache/RemovalCause.html

Reply:
yes, in the existing implementation entry would be removed only in the following cases EXPIRED and SIZE. In these scenarios we would want the eviction behavior. COLLECTED is not relevant, since we are not using WeakKeys/WeakValues/SoftValues. When explicit writeclose is implemented we would do EXPLICIT removal by calling invalidate method explicitly, and even in that case we would need same behavior. In our implementation we would never get into REPLACED scenario, because we are not calling 'put' or 'putAll' method anywhere, value of the entry (EntryLogAndLockTuple) is CacheLoaded by the CacheLoader.

Being said that, I should rename this method and remove word 'Expiry' from name and update the method description. Thanks for pointing that out.